### PR TITLE
Extract JSON credentials

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context
HTTP providers use the same JSON credential format as IMDS

## Description
Straightforward refactoring to extract JSON credentials into a separate module

## Testing
- [x] covered by unit tests

## Checklist
- no changelog

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws/rust-runtime/aws-config/src/json_credentials.rs [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+        match self {
+            InvalidJsonCredentials::Other(msg) => write!(f, "{}", msg),
+            InvalidJsonCredentials::MissingField(field) => write!(
+                f,
+                "Expected field `{}` in IMDS response but it was missing",
+                field
+            ),
+            InvalidJsonCredentials::JsonError(json) => {
+                write!(f, "invalid JSON in IMDS response: {}", json)
+            }
```
Comment (Velfi) [inline aws/rust-runtime/aws-config/src/json_credentials.rs:44]: _nitpick_
Is it conventional to reverse the order these are defined in? I think when we write match statements for enums that the variants order should be the same as when we declared it. Just a style preference.

**Review by Velfi — APPROVED** [review]

File: aws/rust-runtime/aws-config/src/json_credentials.rs [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+       //     // and also: DurationSeconds, ExternalId, SerialNumber, TokenCode, Policy
+       //     ...
+       //   }
+}
+
+/// Deserialize an IMDS response from a string
+///
+/// There are two levels of error here: the top level distinguishes between a successfully parsed
+/// response from IMDS vs. something invalid / unexpected. The inner error distinguishes between
+/// a successful request to IMDS that includes credentials vs. an error with a code.
```
Comment (jdisanti) [inline aws/rust-runtime/aws-config/src/json_credentials.rs:89]: Comment needs to be updated

File: aws/rust-runtime/aws-config/src/json_credentials.rs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+                secret_access_key: "xjtest".into(),
+                session_token: "IQote///test".into(),
+                expiration: UNIX_EPOCH + Duration::from_secs(1631935916),
+            }
+        )
+    }
+
+    /*
+    #[test]
+    fn json_credentials_optional_session_token() {
```
Comment (jdisanti) [inline aws/rust-runtime/aws-config/src/json_credentials.rs:252]: Why is this test commented now?

File: aws/rust-runtime/aws-config/src/json_credentials.rs (hunk 1 above)
Comment (rcoh) [reply aws/rust-runtime/aws-config/src/json_credentials.rs:44]: oh yeah, I think that was just some copy paste evolution

Comment (rcoh) [reply aws/rust-runtime/aws-config/src/json_credentials.rs:44]: fixed

File: aws/rust-runtime/aws-config/src/json_credentials.rs (hunk 3 above)
Comment (rcoh) [reply aws/rust-runtime/aws-config/src/json_credentials.rs:252]: Updated, AFAICT from the spec, token isn't actually optional

**Pull request merged**